fairscale/nn/data_parallel/offload_ddp.py

```diff
@@ -0,0 +1,402 @@
+                self.model_shard.buffers(),
+            ),
+        )
+        return requests if non_blocking else self._sync(requests)
+
+    def sync_parameters(self, non_blocking: bool = True) -> Optional[List[Any]]:
+        """
+        Sync all the parameters in between ranks.
+        """
+        requests = list(
+            map(
+                lambda x: dist.broadcast(x.data, self.owner_rank, self.process_group, async_op=True),
+                self.model_shard.parameters(),
+            ),
+        )
+        return requests if non_blocking else self._sync(requests)
+
+    @staticmethod
+    def _sync(requests: Optional[List[Any]]) -> None:
+        """
+        Make an async function synchronous.
+        Use this to wrap the function call directly
+        """
+        if requests:
+            _ = list(map(lambda x: x.wait(), requests))
+        return
+
+
+class ShardSyncLayer(torch.autograd.Function):
+    """
+     The shard sync layer is a synchronization point between model shards.
+
+     - In the forward pass, it drops parameters in the previous shard and
+     loads parameters for the next shard.
+
+     - In the backward pass, it does the reverse and also gathers gradients to the owner.
+
+     It does not change or create any outputs at all, instead it just
+     forwards the input as the output.
+
+     NOTE: see https://pytorch.org/docs/stable/autograd.html#torch.autograd.Function
+     """
+
+    @staticmethod
+    def forward(ctx: Any, p2_shard: Optional[ModelShard], p1_shard: Optional[ModelShard], n1_shard: Optional[ModelShard], n2_shard: Optional[ModelShard], *inputs: Any) -> Any:  # type: ignore
+        # Drop the shard we just went through, except if this is the last one in line
+        if p1_shard and n1_shard:
+            p1_shard.forward_drop(non_blocking=True)
+
+        # Start the load of the next shard in line, opportunistically look ahead
+        if n2_shard:
+            n2_shard.forward_load(sync=True, non_blocking=True)
+
+        ctx.p2_shard = p2_shard
+        ctx.p1_shard = p1_shard
+        ctx.n1_shard = n1_shard
+        ctx.n2_shard = n2_shard
+
+        outputs = inputs
+
+        return outputs
+
+    @staticmethod
+    def backward(ctx, *grad_outputs):  # type: ignore
+        if ctx.n1_shard:
+            ctx.n1_shard.reduce_grads(non_blocking=True)
+            ctx.n1_shard.backward_drop(non_blocking=True)
+
+        # Opportunistically pre-load ahead of the compute wavefront
+        if ctx.p2_shard:
+            ctx.p2_shard.backward_load(non_blocking=True)
+
+        # The returned variables need to mirror the forward inputs
+        if isinstance(grad_outputs, tuple):
+            return None, None, None, None, grad_outputs[0]
+
+        return None, None, None, None, grad_outputs
+
+
+class OffloadDataParallelExperimental(nn.Module):
+    """Implements distributed data parallel training with optimizer state sharding and model sharding.
```

Comment:
@min-xu-ai the gist of it is here. In this implementation:
- the model is split into sequential shards
- the shards are moved to "offload_device" and loaded to gpu on the fly, depending on where the compute is in the FW or BW pass
- the shard are sync'ed on FW (to get the updated shard)
- same as OSS, just implemented differently, each rank owns the update for only a shard of the model
- same as ShardedDDP, all the other gradients can be discarded after the reduce
- right now the motion is CPU/GPU boundary, but it could be over the network, does not really change the principle